cpp: Upgrade to 2016-DEV0 model

rleigh-dundee opened the pull request:
Add units, quantities, map annotations, folders to the generator, templates etc.
Also includes a number of minor bugfixes.

Note this is based upon https://github.com/openmicroscopy/bioformats/pull/2427.

Source changes: https://ci.openmicroscopy.org/job/BIOFORMATS-DEV-merge-generated-sources/250/

I wonder if this PR breaks any records, at a total 58058 lines of changes in 7919 static and 50141 generated (22856 model, 35202 metadata).  While generated from just 4901 lines of already overly verbose XML schema, the information density in this code is very low.

rleigh-dundee commented:
Green in https://ci.openmicroscopy.org/view/Files-DEV/job/OME-FILES-CPP-DEV-breaking/103/

rleigh-dundee commented:
Note: Rebased onto the latest https://github.com/openmicroscopy/bioformats/pull/2427 to make the code generation look sensible, as well as the history.

dgault commented on components/formats-gpl/utils/writeMapAnnotationsExample.java:
```diff
@@ -39,6 +39,7 @@
 import loci.common.services.ServiceFactory;
 import loci.formats.meta.IMetadata;
 
+import ome.xml.model.*;
```
The import model.* could be replaced with import ome.xml.model.MapPair;

rleigh-dundee replied:
Should be fixed now.

dgault commented:
I know they were commits that I mad, but just looking at the removal of abstractproprietry from the code gen and wanting to clarify, are we sure that it is ok if no longer support the building of previous model versions? (Of course we still support the import of older model files) Is there any reason we would ever want to build those older model objects from the latest release?

dgault commented on components/xsd-fu/python/ome/modeltools/property.py:
```diff
@@ -415,6 +432,17 @@ def _get_isReference(self):
+    concreteClasses = property(
+        _get_concreteClasses,
+        doc="""Concrete instance types for an abstract type.""")
+
```
The get_concreteClasses function must have snuck in with one of my commits to revert to an older schema, it isnt being used and can be removed

dgault commented on components/xsd-fu/templates-cpp/DummyMetadata.template:
```diff
@@ -466,54 +480,28 @@ namespace ome
         // Documented in base class.
         index_type
-        get${abstractClass}Count(${indexes_string(v)}) const;
+        get${abstractClass}Count(${indexes_string(v[:-1])}) const;
```
Should this be indexes_string_dummy also?

rleigh-dundee replied:
Never for headers, only for sources where we don't use the arguments.  The sole reason for doing this is to avoid compiler warnings about unused arguments.  They all look consistent from that point of view.

dgault commented on components/xsd-fu/templates-cpp/OMEXMLMetadata.template:
```diff
@@ -1,4 +1,5 @@
 {% def indexes_string(indexes) %}${", ".join(map(lambda x: lang.index_string(x),indexes))}{% end %}\
+{% def indexes_string_dummy(indexes) %}${", ".join(map(lambda x: lang.index_string(x, dummy=True),indexes))}{% end %}\
```
As the indexes_string_dummy is not being used in this template any more (had previously been in by mistake by the looks of it), the definition can probably be removed

rleigh-dundee replied:
Now dropped, thanks.

dgault commented on components/xsd-fu/templates-cpp/OMEXMLModelObject.template:
```diff
@@ -1,12 +1,3 @@
-      for o in sorted(model.objects.values(), lambda x, y: cmp(x.name, y.name)):
-          if o.parentName == abstractClass:
-            returnList.append(o.name)  
-    return returnList
```
I had made a previous comment about removing getConcreteClasses from property, but now I see you've moved it to there from the template. In which case both changes are fine to leave as they are

dgault commented on components/xsd-fu/templates-cpp/OMEXMLModelObject.template:
```diff
@@ -396,7 +391,7 @@ ${customContent}\
           this->impl->value = text;
 {% end %}\
-        }
+        } //a
```
Rouge comment snuck in

dgault commented on components/xsd-fu/templates-cpp/OMEXMLModelObject.template:
```diff
@@ -480,11 +510,12 @@ ${customUpdatePropertyContent[prop.name]}
+{% end choose %}\
 {% end %}\
-          }
+          } // c
```
2 more 'b' and 'c' comments

dgault commented on components/xsd-fu/templates-cpp/OMEXMLModelObject.template:
```diff
@@ -994,7 +1071,7 @@ ${customUpdatePropertyContent[prop.name]}
 {% end source %}\
 {% end %}\
-{% when prop.maxOccurs > 1 and not klass.isAbstract and not klass.isAbstractProprietary %}
+{% when prop.maxOccurs > 1 and not klass.isAbstract %}\
```
In the cpp generated model code there are a number of places where new lines have disappeared. 
I cant see anywhere particular in the template that removes them but perhaps adding the forward slash (correctly) onto these template conditionals means that the set functions above for example no longer have any spacing between them.

rleigh-dundee replied:
I'll look over this in more detail tomorrow.  As a general rule, every conditional uses `%}\` to end every line to avoid any trailing characters contributing to the output, including newlines, so that every line in the output is explicit and obvious.  This was due to some subtle problems in the past.

dgault commented:
In the cpp model generated code there are a number of empty conditions such as in update functions:
if (element.hasAttribute("ExposureTimeUnit"))
{         
}

Do we need extra code to deal with the units attributes in these cases?

dgault commented:
Finally finished reviewing all the changes here. I've covered all of the code changes, with a number of comments as above. I've also reviewed the generated source changes for both the model and metadata on C++ and Java.

The important thing here is that everything builds and all the tests are good on both platforms. The metadata API changes also appear to be as expected. As the C++ API is moving up 2 schemas the changes here are significant but there is nothing that raises concern from my side. That being said with such significant API updates it would always be good to have a few more eyes scan the cpp metadata diffs.

The only real potential discussion point is the previously mentioned question about supporting the building of older schemas from the latest code base.

rleigh-dundee commented:
@dgault 

```
if (element.hasAttribute("ExposureTimeUnit"))
```

These should now be dropped (not generated).

Regarding the 2013-06 model, I don't think we will want to ever generate it again.  The intent is to only support 2016 here; 2015 was previously the goal but since it required schema changes it wasn't possible to achieve that, so we switched to 2016 instead.  I've cleaned up some of the 2013 support in the templates; if any other support is lingering which wasn't reverted by your changes or mine, it can be deleted as well.

I'm still working though the other comments you made; hopefully I'll get to them early tomorrow.

rleigh-dundee commented:
The latest commit fixes the problem noted by @sbesson in https://github.com/openmicroscopy/bioformats/pull/2427#issuecomment-223927569

rleigh-dundee commented:
That's it for template tidying for this PR I think.  There are some whitespace issues in the metadata templates, but it's hard to fix when the total generated code delta is ~95klines, 60klines of which are metadata code.  It would probably be more prudent to fix as a followup where the diffs would be readable.

I think I've addressed all the points raised so far, unless I missed something.  Any further feedback welcome.

dgault commented:
The updated changes look good to me. I agree that a separate PR to fully tidy up and clean whitespace issues in the templates is probably the best approach.

sbesson commented:
All CI jobs have been green in the last few days and successfully test our set of OME-XML samples including the new ones added in https://github.com/openmicroscopy/bioformats/pull/2447. I can confirm the Java generated sources diff is now fixed and nothing in the large C++ generated sources diff raised immediate concerns as well. Merging this to bring  C++ support for the new model. Congratulations to everyone involved in the last few months especially @rleigh-dundee and @dgault.

Do we have a follow-up card to tidy up the whitespaces in the templates? Actually on the Java side, we might use this opportunity to comply with the indentation style by using spaces instead of tabs.
